File: src/main/java/com/mpe85/grampa/rule/impl/CodePointPredicateRule.java
package com.mpe85.grampa.rule.impl;

import static com.google.common.base.Preconditions.checkNotNull;

import java.util.Objects;
import java.util.function.Predicate;

import com.google.common.base.MoreObjects.ToStringHelper;
import com.ibm.icu.lang.UCharacter;
import com.mpe85.grampa.rule.RuleContext;

/**
 * A code point predicate rule implementation.
 * 
 * @author mpe85
 *
 * @param <T>
 *            the type of the stack elements
 */
public class CodePointPredicateRule<T> extends AbstractRule<T> {
	
	/**
	 * C'tor. Create a code point predicate rules that exactly matches a specific code point.
	 * 
	 * @param codePoint
	 *            a code point
	 */
	public CodePointPredicateRule(final int codePoint) {
		this(cp -> cp == codePoint);
	}
	
	/**
	 * C'tor. Create a code point predicate rule.
	 * 
	 * @param predicate
	 *            a predicate that is tested by the rule.
	 */
	public CodePointPredicateRule(final Predicate<Integer> predicate) {
		this.predicate = checkNotNull(predicate, "A 'predicate' must not be null.");
	}
	
	@Override
	public boolean match(final RuleContext<T> context) {
		return !context.isAtEndOfInput()
				&& predicate.test(context.getCurrentCodePoint())
				&& context.advanceIndex(UCharacter.charCount(context.getCurrentCodePoint()));
	}
	
	@Override
	public int hashCode() {
		return Objects.hash(super.hashCode(), predicate);
	}
	
	@Override
	public boolean equals(final Object obj) {
		if (obj != null && getClass() == obj.getClass()) {
			final CodePointPredicateRule<?> other = (CodePointPredicateRule<?>) obj;
			return super.equals(other)
					&& Objects.equals(predicate, other.predicate);
		}
		return false;
	}
	
	@Override
	protected ToStringHelper toStringHelper() {
		return super.toStringHelper()
				.add("predicate", predicate);
	}
	
	
	private final Predicate<Integer> predicate;
	
}

File: src/main/kotlin/com/mpe85/grampa/rule/impl/CodePointPredicateRule.kt
package com.mpe85.grampa.rule.impl

import com.ibm.icu.lang.UCharacter.charCount
import com.ibm.icu.lang.UCharacter.foldCase
import com.mpe85.grampa.context.ParserContext
import com.mpe85.grampa.rule.Rule
import com.mpe85.grampa.util.checkEquality
import com.mpe85.grampa.util.stringify
import java.util.Objects.hash

/**
 * A code point predicate rule implementation.
 *
 * @author mpe85
 * @param[T] The type of the stack elements
 * @property[predicate] A predicate that is tested by the rule
 */
public open class CodePointPredicateRule<T>(private val predicate: (Int) -> Boolean) : AbstractRule<T>() {

    /**
     * Construct a code point predicate rule that exactly matches a specific code point.
     *
     * @param[codePoint] A code point
     */
    public constructor(codePoint: Int) : this({ it == codePoint })

    override fun match(context: ParserContext<T>): Boolean = !context.atEndOfInput &&
            predicate(context.currentCodePoint) &&
            context.advanceIndex(charCount(context.currentCodePoint))

    override fun hashCode(): Int = hash(super.hashCode(), predicate)
    override fun equals(other: Any?): Boolean = checkEquality(other, { super.equals(other) }, { it.predicate })
    override fun toString(): String = stringify("predicate" to predicate)
}

/**
 * An ignore-case code point rule implementation.
 *
 * @author mpe85
 * @param[T] The type of the stack elements
 * @param[codePoint] A code point
 */
public class IgnoreCaseCodePointRule<T>(codePoint: Int) :
    CodePointPredicateRule<T>({ foldCase(it, true) == foldCase(codePoint, true) })

/**
 * Create a rule from this code point.
 *
 * @return A [CodePointPredicateRule]
 */
public fun <T> Int.toRule(): Rule<T> = CodePointPredicateRule(this)

/**
 * Create an ignore-case rule from this code point.
 *
 * @return A [IgnoreCaseCodePointRule]
 */
public fun <T> Int.toIgnoreCaseRule(): Rule<T> = IgnoreCaseCodePointRule(this)
